crates/bitwarden/src/client/auth_settings.rs

```diff
@@ -2,46 +2,83 @@ use std::num::NonZeroU32;
 
 use base64::Engine;
 use bitwarden_api_identity::models::{KdfType, PreloginResponseModel};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
 
 use crate::{
     crypto::{PbkdfSha256Hmac, PBKDF_SHA256_HMAC_OUT_SIZE},
-    util::{default_kdf_iterations, BASE64_ENGINE},
+    util::{
+        default_argon2_iterations, default_argon2_memory, default_argon2_parallelism,
+        default_pbkdf2_iterations, BASE64_ENGINE,
+    },
 };
 
-#[derive(Debug)]
-pub(crate) struct AuthSettings {
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct AuthSettings {
     pub email: String,
-    kdf_type: KdfType,
-    pub(crate) kdf_iterations: NonZeroU32,
+    #[serde(flatten)]
+    pub(crate) kdf: Kdf,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub enum Kdf {
+    PBKDF2 {
+        iterations: NonZeroU32,
+    },
+    Argon2id {
+        iterations: NonZeroU32,
+        memory: NonZeroU32,
+        parallelism: NonZeroU32,
+    },
 }
 
 impl AuthSettings {
     pub fn new(response: PreloginResponseModel, email: String) -> Self {
-        Self {
-            email,
-            kdf_type: response.kdf.unwrap_or_default(),
-            kdf_iterations: response
-                .kdf_iterations
-                .and_then(|e| NonZeroU32::new(e as u32))
-                .unwrap_or_else(default_kdf_iterations),
-        }
+        let kdf = match response.kdf.unwrap_or_default() {
+            KdfType::Variant0 => Kdf::PBKDF2 {
+                iterations: response
+                    .kdf_iterations
+                    .and_then(|e| NonZeroU32::new(e as u32))
+                    .unwrap_or_else(default_pbkdf2_iterations),
+            },
+            KdfType::Variant1 => Kdf::Argon2id {
+                iterations: response
+                    .kdf_iterations
+                    .and_then(|e| NonZeroU32::new(e as u32))
+                    .unwrap_or_else(default_argon2_iterations),
+
+                memory: response
+                    .kdf_memory
+                    .and_then(|e| NonZeroU32::new(e as u32))
+                    .unwrap_or_else(default_argon2_memory),
+                parallelism: response
+                    .kdf_parallelism
+                    .and_then(|e| NonZeroU32::new(e as u32))
+                    .unwrap_or_else(default_argon2_parallelism),
+            },
```

Comment:
Can we move the argon2 work to a separate PR?

Reply:
Removed from here, PR in progress now